pkg/sql/alter_primary_key.go

```diff
@@ -627,30 +627,43 @@ func (p *planner) shouldCreateIndexes(
 // * The table has a primary key (no DROP PRIMARY KEY statements have
 //   been executed).
 // * The primary key is not the default rowid primary key.
-// * The new primary key isn't the same hash sharded old primary key with a
-//   different bucket count.
+// * The new primary key isn't the same set of columns and directions
+//   other than hash sharding.
 // * There is no partitioning change.
 func shouldCopyPrimaryKey(
 	desc *tabledesc.Mutable,
 	newPK *descpb.IndexDescriptor,
 	alterPrimaryKeyLocalitySwap *alterPrimaryKeyLocalitySwap,
 ) bool {
-	if alterPrimaryKeyLocalitySwap != nil {
-		return false
-	}
-	oldPK := desc.GetPrimaryIndex()
-	if !desc.HasPrimaryKey() {
-		return false
-	}
-	if desc.IsPrimaryIndexDefaultRowID() {
-		return false
+
+	columnIDsAndDirsWithoutSharded := func(idx *descpb.IndexDescriptor) (
+		columnIDs descpb.ColumnIDs,
+		columnDirs []descpb.IndexDescriptor_Direction,
+	) {
+		for i, colName := range idx.KeyColumnNames {
+			if colName != idx.Sharded.Name {
```

Comment:
maybe `columnIDs = columnsIDs[idx. ExplicitColumnStartIdx():]`, so it takes care of `crdb_region` as well?